File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {

        List<List<Integer>> traversal = new ArrayList<>();

        if (root == null) {
            return traversal;
        }

        Queue<TreeNode> q = new LinkedList<>();

        TreeNode delim = new TreeNode(-1);

        q.add(root);
        q.add(delim);

        List<Integer> curr = new ArrayList<>();

        while(!q.isEmpty()) {
            TreeNode node = q.poll();

            if (node == delim) {
                if (!q.isEmpty()) {
                    q.add(delim);
                }
                List<Integer> temp = new ArrayList<>(curr);
                traversal.add(temp);
                curr = new ArrayList<>();
            } else {
                curr.add(node.val);

                if (node.left != null) {
                    q.add(node.left);
                }

                if (node.right != null) {
                    q.add(node.right);
                }
            }
        }

        return traversal;
    }
}

File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []

        self.go(n, res, "", 0, 0)

        return res
    
    def go(self, n, res, curr, open_, close):
        if open_ > n:
            return
        
        if close > open_ or close > n:
            return
        
        if open_ == close and open_ == n and close == n:
            res.append(curr[:])
            return
        
        # two conditions: left or right
        self.go(n, res, curr + '(', open_ + 1, close)
        self.go(n, res, curr + ')', open_, close + 1)


        